### ISSUE TITLE
Product catalog pages are blank when CacheAPC is enabled

### ISSUE THREAD

**Issue by AtteR** [PrestaShop/PrestaShop#16245]
Hello,

We are having some problems with CacheAPC. First issues were that when enabled, after some time the webstore stopped completely loading, but this was fixed by increasing memory for APC. After increasing memory store seem to be working fine, except some product catalog pages were broken now with following error:

PHP Warning:  array_merge(): Argument #1 is not an array in .../classes/ProductAssembler.php on line 95
PHP Fatal error:  Uncaught TypeError: Argument 2 passed to PrestaShop\PrestaShop\Adapter\Presenter\Product\ProductListingPresenter::present() must be of the type array, boolean given

Weird thing with these crashes it only happens in our release environment, not in our testing enviroment (probably has to do something with user count being higher in release env), so tracking down the crashes is a bit hard. I am not sure if this product catalog bug has something to do with APC module or is it bug inside Prestashop?

PrestaShop version: 1.7.6.1
PHP version: 7.2.22

**Comment by khouloudbelguith** [comment]
Hi @AtteR,

Thanks for your bug report.
I did not manage to reproduce the issue with PS1.7.6.1 & APCu v5.1.18.
Here is a link to Optimize your PrestaShop performance.
There is a Prestashop module that is promising for APCu.
https://dh42.com/free-prestashop-modules/apcu-manager/
This module just displays the APCu dashboard and doesn't have any impact on the configuration or performances.
I attached a screen record
https://drive.google.com/file/d/1AWYCg0OEH-KNXTlVNex-LzX8Rot5SOHd/view
Thanks to check & feedback.

**Comment by AtteR** [comment]
Hi,

Sorry, I meant the webstore frontend product catalog pages or category pages (not sure which one they are called). I should have been more specific.

We have installed and tracked the caching with APCu info (https://github.com/krakjoe/apcu/blob/master/apc.php). There is plenty memory available but the fragmentation is 100% after hour or so. I am not sure if the issue is because of fragmentation. Is there any config recommendations for Prestashop? Below is our current APCu configs.

![apcu configs](https://user-images.githubusercontent.com/5371194/68022012-adfa9b00-fcab-11e9-9cbe-2398c45ed470.JPG)

**Comment by khouloudbelguith** [comment]
@AtteR, thanks for your feedback.
No, I tried with PS1.7.6.1 also in the Front office & it is well working.
I attached a screen record
https://drive.google.com/file/d/1JX4oYdag0OXVpnesTj3O9X5Mm2j7rKAA/view
Thanks!

**Comment by AtteR** [comment]
Hi,

Thanks for the videos.

Our webstore has a lot of different categories and products (thousands). I was able to get test enviroment to crash also. APCu fragmentation is 100% and there is still some memory left, but some product categories currently crash with this error:

Warning: array_merge(): Argument #1 is not an array
in ProductAssembler.php line 95

If your apc.ttl is 0, the apc will clear all the cache when there is no space left, which also clears fragmentation. If this is set to 3600 for example, and you visit a lot of different product category pages, at some point category pages start crashing. I am still unsure if this simply misconfigurated APC or is there also bug somewhere. I will look into the error message and see if I can fix it from code.

**Comment by khouloudbelguith** [comment]
@AtteR, in my case it is also 3600
![image](https://user-images.githubusercontent.com/16067358/68029616-e3a58100-fcb7-11e9-85a0-07bbc719bd25.png)
could you please provide the whole exception displayed attached in PDF file.

Thanks!

**Comment by AtteR** [comment]
Hello,

Here is a screenshot, I could not post PDF here.

![prestashop apc crash](https://user-images.githubusercontent.com/5371194/68029978-143de880-fcc1-11e9-966f-b49fbe86a696.png)

**Comment by khouloudbelguith** [comment]
@AtteR, did you delete a product which has an ID=20141950?

Did you use the module ps_viewedproduct? what is the exact version?

Thanks!

**Comment by AtteR** [comment]
Yes, actually there is no product with ID = 2014950 in ps_product table.

ps_viewedproduct is enabled and the version is 1.2.1.

**Comment by khouloudbelguith** [comment]
@AtteR, could you please check your Database table ps_product, if this ID exists?

Thanks!

**Comment by AtteR** [comment]
SELECT * FROM ps_product WHERE id_product = 2014950; return zero rows, so it does not exist.

**Comment by khouloudbelguith** [comment]
@AtteR, sorry it is 20141950.
![image](https://user-images.githubusercontent.com/16067358/68031560-384afb00-fcbc-11e9-86dc-0a4b22ebce04.png)

Thanks!

**Comment by AtteR** [comment]
Oh, sorry yes.

SELECT * FROM ps_product WHERE id_product = 20141950;

Finds 1 row.

**Comment by khouloudbelguith** [comment]
@AtteR, this product it is displayed in your BO?

Thanks!

**Comment by AtteR** [comment]
No, it is missing in BO.

**Comment by AtteR** [comment]
Yes, actually it is there.

**Comment by khouloudbelguith** [comment]
@AtteR, could you please try to disable it, clear cache, check & feedback.

Thanks!

**Comment by AtteR** [comment]
Clearing cache does fix it, but then after sometime some other page will break.

**Comment by khouloudbelguith** [comment]
@AtteR, the same product is displayed in your exception?

Thanks!

**Comment by AtteR** [comment]
Different product now in the exception.

![Screenshot_2019-11-01 Screenshot](https://user-images.githubusercontent.com/5371194/68032456-65041000-fcc6-11e9-9b33-f1cfd5da3af9.png)

**Comment by khouloudbelguith** [comment]
@AtteR, if you disable the cacheAPC => it is OK?

Thanks!

**Comment by AtteR** [comment]
Yes, I disabled CacheAPC without clearing the cache and the product category page works.

**Comment by khouloudbelguith** [comment]
@AtteR, so, it is a cache issue.
Thanks to follow these steps:

- `Never recompile template files`: this option is to be chosen for maximum performance.
- `Cache enabled (YES)`: makes pages fastest. 
- `Smart cache for CSS (YES)`: brings together in a single file, compresses and caches CSS files.
- `Smart cache for Javascript (YES)`: brings together in a single file, compresses and caches JS files.
- `Apache Optimization (YES)`: allows to further increase site performance.

**Comment by AtteR** [comment]
Hi,

Yes, it is cache issue. We do have settings enabled like that, but what we should do with APCu?

**Comment by khouloudbelguith** [comment]
Hi @AtteR,

I'm trying to create a huge shop with a lot of products & check the issue again.
Because it is OK with a normal shop.

Thanks!

**Comment by khouloudbelguith** [comment]
Hi @AtteR,

I tried with a huge shop with a lot of products & it is OK.
CacheAPC is enabled => there is no issue in the FO.
I tried with PS1.7.6.1 & PS1.7.5.2 => OK.
https://drive.google.com/file/d/1V0xdmJYlH9-McYq2JdW3DxAXmJHuMeos/view
Despite our several trials, we could not reproduce your issue with the provided information.
It seems that your issue is not a PrestaShop's core bug but most likely a server configuration or customization problem.
I invite you to contact our support team: http://addons.prestashop.com/en/388-support to help you solve your issue.

Thanks!
